Source/DafnyLanguageServer.Test/Various/ShutdownTest.cs

```diff
@@ -0,0 +1,182 @@
+      var process = await StartLanguageServerRunnerProcess();
+      var error = "";
+      process.ErrorDataReceived += delegate (object sender, DataReceivedEventArgs args) { error += args.Data; };
+
+      var languageServerProcessId = await process.StandardOutput.ReadLineAsync();
+      Assert.IsNotNull(languageServerProcessId);
+
+      Thread.Sleep(1000); // Give the process some time to die
+      var initializeMessage = GetLspInitializeMessage(null);
+      await process.StandardInput.WriteAsync(initializeMessage);
+
+      var initializedResponseFirstLine = await process.StandardOutput.ReadLineAsync();
+      Assert.IsFalse(string.IsNullOrEmpty(initializedResponseFirstLine));
+
+      var didExit = process.WaitForExit(-1);
+      Assert.IsTrue(didExit);
+
+      Thread.Sleep(100); // Give the process some time to die
+
+      Assert.IsFalse(string.IsNullOrEmpty(initializedResponseFirstLine));
+      try {
+        var languageServer = Process.GetProcessById(int.Parse(languageServerProcessId));
+        languageServer.Kill();
+      } catch (ArgumentException) {
+        Assert.Fail("Language server should not have killed itself if it doesn't know the parent.");
+      }
+    }
+
+    [TestMethod]
+    public async Task LanguageServerShutsDownIfParentDies() {
+      var process = await StartLanguageServerRunnerProcess();
+
+      var languageServerProcessId = await process.StandardOutput.ReadLineAsync();
+
+      var initializeMessage = GetLspInitializeMessage(process.Id);
+      await process.StandardInput.WriteAsync(initializeMessage);
+
+      var initializedResponseFirstLine = await process.StandardOutput.ReadLineAsync();
+      Assert.IsFalse(string.IsNullOrEmpty(initializedResponseFirstLine));
+
+      var error = await process.StandardError.ReadToEndAsync();
+      var didExit = process.WaitForExit(-1);
+      Assert.IsTrue(didExit);
+      Assert.IsNotNull(languageServerProcessId, error);
+      Assert.IsTrue(string.IsNullOrEmpty(error), error);
+
+      // Wait for the language server to kill itself by waiting until it closes the output stream.
+      await process.StandardOutput.ReadToEndAsync();
+      Thread.Sleep(100); // Give the process some time to die
+      try {
+        var languageServer = Process.GetProcessById(int.Parse(languageServerProcessId));
+        languageServer.Kill();
+        Assert.Fail("Language server should have killed itself if the parent is gone.");
+      } catch (ArgumentException) {
+        // Language server process is not running, pass the test.
+      }
+    }
+
+    private static async Task<Process> StartLanguageServerRunnerProcess() {
+      var languageServerBinary = Path.Join(Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location), "DafnyLanguageServer");
+      var languageServerRunnerPath = await CreateDotNetDllThatStartsGivenFilepath(languageServerBinary);
+
+      var processInfo = new ProcessStartInfo("dotnet", languageServerRunnerPath) {
```

Comment:
Why do we run `dotnet DafnyLanguageServer.dll` instead of `DafnyLanguageServer.exe` directly?

Reply:
Oh, I see what's happening, sorry.